src/wallet/test/coinselector_tests.cpp

```diff
@@ -917,5 +917,176 @@ BOOST_AUTO_TEST_CASE(effective_value_test)
     BOOST_CHECK_EQUAL(output5.GetEffectiveValue(), nValue); // The effective value should be equal to the absolute value if input_bytes is -1
 }
 
+BOOST_AUTO_TEST_CASE(check_max_weight)
+{
+    const CAmount target = 49.5L * COIN;
+    CCoinControl cc;
+
+    std::unique_ptr<CWallet> tmpWallet = std::make_unique<CWallet>(m_node.chain.get(), "", m_args, CreateMockWalletDatabase());
+    tmpWallet->LoadWallet();
+    LOCK(tmpWallet->cs_wallet);
+    tmpWallet->SetupLegacyScriptPubKeyMan();
+
+    const auto op_dest = tmpWallet->GetNewDestination(OutputType::BECH32, "");
+    BOOST_ASSERT(op_dest);
+    const auto scriptPubKey = GetScriptForDestination(op_dest.value());
+    const auto txout = CTxOut(target, scriptPubKey);
+
+    FastRandomContext rand;
+    CoinSelectionParams cs_params{
+        rand,
+        /*change_output_size=*/34,
+        /*change_spend_size=*/148,
+        /*min_change_target=*/CENT,
+        /*effective_feerate=*/CFeeRate(0),
+        /*long_term_feerate=*/CFeeRate(0),
+        /*discard_feerate=*/CFeeRate(0),
+        /*tx_noinputs_size=*/static_cast<size_t>(10) + GetSerializeSize(txout, PROTOCOL_VERSION), // static header size + output size
+        /*avoid_partial=*/false,
+    };
+
+    CMutableTransaction txDummy;
+    txDummy.vout.push_back(txout);
+
+    {
+        // Scenario 1:
```

Comment:
ISTM these tests could be deduplicated intoa helper function with each scenario just being a set of parameters as each test case does basically the same thing but with slightly different params.